certora/AllocatorVault.spec

```diff
@@ -0,0 +1,318 @@
+    mathint rate = vat.rate() + (jug.duty() - RAY()) * (e.block.timestamp - jug.rho());
+    require rate > 0;
+    mathint dart = wad * RAY() / rate;
+
+    wipe(e, wad);
+
+    mathint wardsAfter = wards(any);
+    address jugAfter = jug();
+    mathint nstTotalSupplyAfter = nst.totalSupply();
+    mathint nstBalanceOfBufferAfter = nst.balanceOf(buffer());
+    mathint vatInkVaultAfter; mathint vatArtVaultAfter;
+    vatInkVaultAfter, vatArtVaultAfter = vat.urns(ilk(), currentContract);
+
+    assert wardsAfter == wardsBefore, "wipe did not keep unchanged every wards[x]";
+    assert jugAfter == jugBefore, "wipe did not keep unchanged jug";
+    assert vatInkVaultAfter == vatInkVaultBefore, "wipe did not keep vat.urns(ilk,vault).ink unchanged";
+    assert vatArtVaultAfter == vatArtVaultBefore - dart, "wipe did not decrease vat.urns(ilk,vault).art by dart";
+    assert nstBalanceOfBufferAfter == nstBalanceOfBufferBefore - wad, "wipe did not decrease nst.balanceOf(buffer) by wad";
+    assert nstTotalSupplyAfter == nstTotalSupplyBefore - wad, "wipe did not decrease nst.totalSupply() by wad";
+}
+
+// Verify revert rules on wipe
+rule wipe_revert(uint256 wad) {
+    env e;
+
+    bool canCall = roles.canCall(ilk(), e.msg.sender, currentContract, to_bytes4(0xb38a1620));
+    mathint wardsSender = wards(e.msg.sender);
+    mathint nstTotalSupply = nst.totalSupply();
+    address buffer = buffer();
+    require buffer != currentContract;
+    mathint nstBalanceOfBuffer = nst.balanceOf(buffer);
+    mathint nstBalanceOfVault = nst.balanceOf(currentContract);
+    require nstBalanceOfBuffer + nstBalanceOfVault <= nstTotalSupply;
+    mathint nstAllowanceBufferVault = nst.allowance(buffer, currentContract);
+    mathint nstAllowanceVaultNstJoin = nst.allowance(currentContract, nstJoin);
+    mathint vatInkVault; mathint vatArtVault;
+    vatInkVault, vatArtVault = vat.urns(ilk(), currentContract);
+    mathint duty = jug.duty();
+    require duty >= RAY();
+    mathint rho = jug.rho();
+    require to_mathint(e.block.timestamp) >= rho;
+    mathint rate = vat.rate() + (duty - RAY()) * (e.block.timestamp - jug.rho());
+    require rate > 0 && rate <= max_int256();
+    mathint dart = wad * RAY() / rate;
+    mathint vatDaiVault = vat.dai(currentContract);
+    mathint vatDaiNstJoin = vat.dai(nstJoin);
+
+    wipe@withrevert(e, wad);
+
+    bool revert1  = e.msg.value > 0;
+    bool revert2  = !canCall && wardsSender != 1;
+    bool revert3  = nstBalanceOfBuffer < to_mathint(wad);
+    bool revert4  = nstAllowanceBufferVault < to_mathint(wad);
+    bool revert5  = wad * RAY() > max_uint256;
+    bool revert6  = nstAllowanceVaultNstJoin < to_mathint(wad);
+    bool revert7  = vatArtVault < dart;
+    bool revert8  = vatDaiNstJoin < wad * RAY();
+    bool revert9  = vatDaiVault + wad * RAY() > max_uint256;
+    bool revert10 = dart > max_int256();
+    bool revert11 = rate * dart > max_int256();
```

Comment:
I see that reverts 10 and 11 correspond to two different reverting instructions in `vat.frob()`, but since we required that `rate > 0`, `revert11` is always implied by `revert10` so does it do anything to keep `revert10` here?

It seems that we could technically get rid of it (I tried running the prover without it and the rule succeeded), though maybe the goal here is to be explicit about all the different revert conditions even when from the prover standpoint some condition may be redundant?

Reply:
Note also that `rate == 0` would cause a revert for a different reason (division by zero when calculating `dart`), which is not captured here. IMO it's fine to omit this redundant revert condition since the proof is specialized to `rate > 0` just like the division-by-zero revert is excluded. Leaving it in could even be considered slightly misleading/confusing.

OTOH one could make the argument that the proof _shouldn't_ be specialized to `rate > 0` to allow capturing the full behavior of the function (or a separate `rate == 0` rule could be used, which would be slightly cleaner). I might do that personally just to make sure all `wipe` code paths are covered.

Reply:
I notice also the `rate * dart` revert was removed from `draw_revert`, presumably for the same reason.

Reply:
Yeah assert `revert10` should be gone. I'll take care of it.
The thing about exploring the `rate == 0` branch is that if I recall correctly the tool was giving division by zero per se. I might be able to avoid taht making a conditional when defining `mathint dart = wad * RAY() / rate;`. Not sure if it is worth it to complicate it though and still not sure if that will solve the issue I was having.

Reply:
Resolving this one.